File: Editorial/49_Group_Anagrams/Go/1_Categorize_by_Sorted_String.go
func groupAnagrams(strs []string) [][]string {
    // Create a map to store the anagrams, where the keys are the sorted strings and the values are the lists of anagrams
    ans := make(map[string][]string)
    for _, s := range strs {
        key := sortString(s)
        // If the string is an anagram of another string, they will have the same key and thus be grouped together
        ans[key] = append(ans[key], s)
    }

    // Build the result slice from the map entries
    var result [][]string
    for _, group := range ans {
        result = append(result, group)
    }
    return result
}

// sortString sorts the characters of a string and returns the sorted string
func sortString(s string) string {
    sorted := strings.Split(s, "")
    sort.Strings(sorted)
    return strings.Join(sorted, "")
}

File: Editorial/49_Group_Anagrams/TypeScript/1_Categorize_by_Sorted_String.ts
function groupAnagrams(strs: string[]): string[][] {
    // Create a map to store the anagrams, where the keys are the sorted strings and the values are the lists of anagrams
    let map = new Map<string, string[]>();
    // Iterate over the given strings
    for (let str of strs) {
        // Convert the string to a char array and sort it
        let chars = Array.from(str);
        chars.sort();
        // Use the sorted string as a key
        // If the string is an anagram of another string, they will have the same key and thus be grouped together
        let key = chars.join("");
        // If the key is not already in the map, add it with a new list as its value
        if (!map.has(key)) map.set(key, []);
        // Add the original string to the list of its key
        map.get(key).push(str);
    }
    // Return the lists of anagrams
    return Array.from(map.values());
}
